src/field_ground_effect.c

```diff
@@ -238,3 +347,659 @@ void GetGroundEffectFlags_JumpLanding(struct MapObject *mapObj, u32 *flags)
+{
+    if (mapObj->mapobj_bit_4)
+        return;
+
+    if (!MetatileBehavior_IsLongGrass(mapObj->mapobj_unk_1E))
+        return;
+
+    if (!MetatileBehavior_IsLongGrass(mapObj->mapobj_unk_1F))
+        return;
+
+    sprite->subspriteTableNum = 4;
+
+    if (ZCoordToPriority(mapObj->elevation) == 1)
+        sprite->subspriteTableNum = 5;
+}
+
+bool8 IsZCoordMismatchAt(u8 z, s16 x, s16 y)
+{
+    u8 mapZ;
+
+    if (z == 0)
+        return FALSE;
+
+    mapZ = MapGridGetZCoordAt(x, y);
+
+    if (mapZ == 0 || mapZ == 0xF)
+        return FALSE;
+
+    if (mapZ != z)
+        return TRUE;
+
+    return FALSE;
+}
+
+void FieldObjectUpdateZCoordAndPriority(struct MapObject *mapObj, struct Sprite *sprite)
+{
+    if (mapObj->mapobj_bit_26)
+        return;
+
+    FieldObjectUpdateZCoord(mapObj);
+
+    sprite->subspriteTableNum = gFieldObjectPriorities_08376070[mapObj->elevation];
+    sprite->oam.priority = gFieldObjectPriorities_08376060[mapObj->elevation];
+}
+
+void InitObjectPriorityByZCoord(struct Sprite *sprite, u8 z)
+{
+    sprite->subspriteTableNum = gFieldObjectPriorities_08376070[z];
+    sprite->oam.priority = gFieldObjectPriorities_08376060[z];
+}
+
+u8 ZCoordToPriority(u8 z)
+{
+    return gFieldObjectPriorities_08376060[z];
+}
+
+void FieldObjectUpdateZCoord(struct MapObject *mapObj)
+{
+    u8 z = MapGridGetZCoordAt(mapObj->coords2.x, mapObj->coords2.y);
+    u8 z2 = MapGridGetZCoordAt(mapObj->coords3.x, mapObj->coords3.y);
+
+    if (z == 0xF || z2 == 0xF)
+        return;
+
+    mapObj->mapobj_unk_0B_0 = z;
+
+    if (z != 0 && z != 0xF)
+        mapObj->elevation = z;
+}
+
+void SetObjectSubpriorityByZCoord(u8 a, struct Sprite *sprite, u8 b)
+{
+    s32 tmp = sprite->centerToCornerVecY;
+    u32 tmpa = *(u16 *)&sprite->pos1.y;
+    u32 tmpb = *(u16 *)&gSpriteCoordOffsetY;
+    s32 tmp2 = (tmpa - tmp) + tmpb;
+    u16 tmp3 = (0x10 - ((((u32)tmp2 + 8) & 0xFF) >> 4)) * 2;
+    sprite->subpriority = tmp3 + gUnknown_08376050[a] + b;
+}
+
+void FieldObjectUpdateSubpriority(struct MapObject *mapObj, struct Sprite *sprite)
+{
+    if (mapObj->mapobj_bit_26)
+        return;
+
+    SetObjectSubpriorityByZCoord(mapObj->elevation, sprite, 1);
+}
+
+bool8 AreZCoordsCompatible(u8 a, u8 b)
+{
+    if (a == 0 || b == 0)
+        return TRUE;
+
+    if (a != b)
+        return FALSE;
+
+    return TRUE;
+}
+
+#define READ8(ptr, s, member) ((ptr)[offsetof(s, member)])
+
+void GroundEffect_SpawnOnTallGrass(struct MapObject *mapObj, struct Sprite *sprite)
+{
+    u8 *ptr;
+
+    gUnknown_0202FF84[0] = mapObj->coords2.x;
+    gUnknown_0202FF84[1] = mapObj->coords2.y;
+    gUnknown_0202FF84[2] = mapObj->elevation;
+    gUnknown_0202FF84[3] = 2;
+    gUnknown_0202FF84[4] = (mapObj->localId << 8) | mapObj->mapNum;
+    gUnknown_0202FF84[5] = mapObj->mapGroup;
+
+    ptr = (u8 *)&gSaveBlock1;
+    gUnknown_0202FF84[6] = READ8(ptr, struct SaveBlock1, location.mapNum) << 8
+                         | READ8(ptr, struct SaveBlock1, location.mapGroup);
```

Comment:
This matches.
`gUnknown_0202FF84[6] = ((u8)gSaveBlock1.location.mapNum << 8) | (u8)gSaveBlock1.location.mapGroup;`